pkg/controller/daemon/daemon_controller.go

```diff
@@ -1202,10 +1282,28 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
 	if err != nil {
 		return err
 	}
+
+	dsNamespacedName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	// If our writes have not yet been observed by our informers, we should not
+	// continue since our reads will not be in sync with our previously enacted
+	// state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.StaleControllerConsistency) && !dsc.consistencyStore.IsReady(dsNamespacedName) {
+		klog.InfoS("Daemonset is not ready for reconciliation, previous writes have not been observed in the cache", "ds", dsNamespacedName)
+		return nil
+	}
+
 	ds, err := dsc.dsLister.DaemonSets(namespace).Get(name)
 	if apierrors.IsNotFound(err) {
 		logger.V(3).Info("Daemon set has been deleted", "daemonset", key)
 		dsc.expectations.DeleteExpectations(logger, key)
+		dsc.consistencyStore.Clear(
+			dsNamespacedName,
+			ds.UID,
+		)
```

Comment:
this sync loop is the only thing that was going to write to this record, and it couldn't find the daemonset, we should unconditionally delete any entry regardless of uid:

```suggestion
		dsc.consistencyStore.Clear(dsNamespacedName, "")
```